### ISSUE TITLE
NamedTuple style Models (immutable instances, _replace method, etc)

### ISSUE THREAD

**Issue by graingert** [samuelcolvin/pydantic#38]
To support immutable data structures there will need to be some slight tweaks to the BaseModel:

you cannot override the constructor, and would not want to, as there should be a type safe way to construct instances. Will need a staticmethod: eg parse.

```python
from datetime import datetime
from typing import NamedTuple
from pydantic import ImmutableModel

class UserModel(ImmutableModel, NamedTuple):
    id: int = ...
    name = 'John Doe'
    signup_ts: datetime = None

external_data = {'id': '123', 'signup_ts': '2017-06-01 12:22'}
user = UserModel.parse(**external_data)
print(user)
# > UserModel id=123 name='John Doe' signup_ts=datetime.datetime(2017, 6, 1, 12, 22)
print(user.id)
# > 123
print(user[0])
# > 123
print(user._replace(name='foo'))
# > UserModel id=123 name='foo' signup_ts=datetime.datetime(2017, 6, 1, 12, 22)
```

**Comment by samuelcolvin** [comment]
can you give some more details.

**Comment by graingert** [comment]
@samuelcolvin added some more details

**Comment by samuelcolvin** [comment]
Not a complete solution, but `_replace` is already available via [setattr](https://github.com/samuelcolvin/pydantic/blob/master/pydantic/main.py#L87).

As the docstring says this was preferable to overriding `__setattr__` as that caused a significant performance decrease.

I'd be happy to consider making models immutable if there's a way which doesn't effect (or improves) performance and doesn't effect the simple & main use case of pydantic: define a class with some fields, initialise it with data. Done.

**Comment by graingert** [comment]
@samuelcolvin _replace is different from setattr because setattr mutates the model in place.

ImmutableModel would be an alternative API available for users who want to program more pedantically.

**Comment by samuelcolvin** [comment]
closed via #44
